proposed/SerializerGoals5.0.md

```diff
@@ -0,0 +1,281 @@
+- Improve caching
+- Other smaller features TBD.
+
+Reflection emit is used both directly and indirectly by the serializer:
+  - Directly: used to generate getters, setters and calls to constructors.
+  - Indirectly: used to support strongly-typed converters for value types that avoid boxing. This includes usages of `Type.MakeGenericType` (and possibly `Type.MakeGenericMethod` pending additional research).
+
+Some early issues created to start tracking the work:
+- [Avoid Type.MakeGenericType when (de)serializing objects](https://github.com/dotnet/runtime/issues/35028)
+- [Add deferred loading of System.Private.Uri assembly by the serializer](https://github.com/dotnet/runtime/issues/35029). Minor startup gains; low-hanging fruit.
+
+## Reflection features
+As described in the [System.Reflection roadmap](https://github.com/dotnet/runtime/issues/31895) it is possible to add new APIs that will eliminate common usages of Reflection.Emit including getters\setter\constructors. This will help with the direct usage of Reflection.Emit, but not the indirect usage.
+
+Note that AOT code-gen options likely eliminate the need for new reflection features for those POCOs that are code-gen'd. However, we can't assume or force all POCOs to be code-gen'd.
+
+Misc other thoughts:
+    - A fallback (not recommended currently) is to change the internal converters and box for value types (and generic collections of value types). This will create many more short-lived allocs and decrease steady-state performance but allow the serializer to at least function in a "limp-mode" (roughly half of the throughput).
+    - Another fallback would apply to a code-gen option where the generated POCO code closes the generic collections at build time. i.e. instantiate (or reference) a `JsonConverter<List<int>>` directly in the generated code instead of relying on the serializer to close `JsonConverter<T>` to `JsonConverter<List<int>>`.
+
+## Converter code-gen
+A prototype was created for this by @layomia.
+
+It used a simple (non-Roslyn) code generator using the existing custom Value converter model. Various serializer features (null handling, property lookup on deserialization, reference handling, async support, use of custom converters, etc) would need to be baked into the generated code. Thus the resuling code generation will be large, complex and not servicable. At the extreme (with no new helper methods exposed the serializer), each converter would need to implement all of the various features of the serializer.
+
+## Metadata provider code-gen
+A run-time only prototype (not the actual generator) was created for this by @steveharter (see sample generated code below). It makes public the existing internal metadata classes including `JsonClassInfo` and `JsonPropertyInfo`.
+
+It allows for minimal code gen and overlaps with other requested features including property ordering, before\after callbacks and programmatic reading\writing of metadata.
```

Comment:
Backslash is not used for punctuation in English. Also, I'd recommend https://docs.microsoft.com/en-us/style-guide/punctuation/slashes:

> Don't use a slash as a substitute for *or*, like *product/service*. If the UI uses a slash in this way, follow the interface in describing the label, but use *or* to describe the action in text.

```suggestion
It allows for minimal code gen and overlaps with other requested features including property ordering, before and after callbacks and programmatic reading and writing of metadata.
```

Reply:
(+ several other locations)